### ISSUE TITLE
How to fit a dataset to hawkes process

### ISSUE THREAD

**Issue by OnlyAnJ** [X-DataInitiative/tick#179]
Now I want to fit a dataset of gold to hawkes process to see some property, but I am not sure to use which new function to fit this data, maybe ModelHawkesExpKernLogLik, but I don't see the example of this function, could you add some examples of this function?

**Comment by Mbompr** [comment]
Hi @OnlyAnJ ,
Fitting a Hawkes process with likelihood loss instead of least-squares (`ModelHawkesLeastSq`) is quite similar. You just need to change the `gofit` parameter in the learner constructor. However, Hawkes likelihood are more difficult to optimize and `step` and `start` point must be tuned more carefully.

Here is a sample code that fits a Hawkes with likelihood loss on generated data. It is directly adapted from [this example](https://x-datainitiative.github.io/tick/auto_examples/plot_hawkes_matrix_exp_kernels.html).

```python

from tick.plot import plot_hawkes_kernels
from tick.hawkes import SimuHawkesExpKernels, SimuHawkesMulti, HawkesExpKern
import matplotlib.pyplot as plt

end_time = 10000
n_realizations = 10

decays = 3.
baseline = [0.12, 0.07]
adjacency = [[.3, 0.], [.6, .21]]

hawkes_exp_kernels = SimuHawkesExpKernels(
    adjacency=adjacency, decays=decays, baseline=baseline,
    end_time=end_time, verbose=False, seed=1039)

multi = SimuHawkesMulti(hawkes_exp_kernels, n_simulations=n_realizations)

multi.end_time = [(i + 1) / 10 * end_time for i in range(n_realizations)]
multi.simulate()

learner = HawkesExpKern(decays, penalty='l1', C=10, gofit='likelihood', 
                        verbose=True, tol=1e-11, solver='svrg', step=1e-3)
learner.fit(multi.timestamps, start=0.1)

plot_hawkes_kernels(learner, hawkes=hawkes_exp_kernels)
```

Otherwise, if you want to use directly `ModelHawkesExpKernLogLik` class, you might want to have a look a the [Solver class API](https://x-datainitiative.github.io/tick/modules/solver.html#the-solver-class-api). The result above can be replicated using `ModelHawkesExpKernLogLik`, `ProxL1` and `SVRG` classes.

Finally, you might also want to have a look at `HawkesSumExpKern` which allow you to deal with more flexible sum-exponential kernels.

Tell us if you have any other question.

**Comment by OnlyAnJ** [comment]
Sorry for late response, I use a long time to adapt the new version. Now what I want to do is to fit one data by using hawkes process, (attach file is the data). I am confused how do you set initial decays, baseline or other coefficients. I just want to calculate the alpha and beta at every timestamp. 
[goldprice.txt](https://github.com/X-DataInitiative/tick/files/1749841/goldprice.txt)

**Comment by Mbompr** [comment]
In my previous post, all lines before `learner = HawkesExpKern(...)` are generating data with our simulation tool `SimuHawkesExpKernels`. 
Since you have your own data you can skip this part but I needed it to give you a fully working example.

In your specific case, you only have one time serie so you will want to fit a 1-dimensional Hawkes process. In my example, I had a 2-dimensional Hawkes process and that's why I had 2 baselines (usually denoted by mu) and 4 adjacency values (usually denoted by alpha). 

In order to run this on your dataset you should have two things in mind: 
* With HawkesExpKernel, the decay (beta) should be given. This is because otherwise, the loss is not convex and much harder to optimize. You can select the right decay by cross-validation or you can try to estimate it with non-parametric methods such as HawkesEM.
* Your data will need to be shifted. The origin of time is zero so you will need to subtract the minimum time to all your times.

I hope it is clearer now, tell me if you have other questions.

**Comment by OnlyAnJ** [comment]
So if I want to fit a 1-dimensional Hawkes process, I need to set 1 baseline and 1 decay by myself, and calculate alpha in every timestamp? And I want to know how to use cross-validation method to select decay, do I need to set training set and test set?

**Comment by Mbompr** [comment]
The baseline will be estimated together with alpha. Only the decay must be given in advance.

Sorry I used the wrong term, It's not really about cross-validation but rather grid searching. The idea is that you can try many values for the decay and look at the one that will output the highest [score](https://x-datainitiative.github.io/tick/modules/generated/tick.hawkes.HawkesExpKern.html#tick.hawkes.HawkesExpKern.score).

**Comment by stephanegaiffas** [comment]
Can you create an issue with the code generating the segfault ?
Thanks

**Comment by akshitbansal** [comment]
Find max log likelihood shows error " score() missing 1 required positional argument:'self' "
